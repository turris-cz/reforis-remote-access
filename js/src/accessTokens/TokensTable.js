/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";
import {
    withErrorMessage,
    withSpinnerOnSending,
    Button,
    DownloadButton,
    useAlert,
    useAPIDelete,
    API_STATE,
    SpinnerElement,
    formFieldsSize,
} from "foris";

import API_URLs from "API";
import TOKEN_STATUS from "./token_status";

const tokenShape = PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    status: PropTypes.string.isRequired,
});

TokensTable.propTypes = {
    tokens: PropTypes.arrayOf(tokenShape).isRequired,
};

function TokensTable({ tokens }) {
    if (!tokens || tokens.length === 0) {
        return (
            <p className="text-muted text-center">
                {_("There are no tokens added yet.")}
            </p>
        );
    }

    return (
        <table className={`table table-hover ${formFieldsSize}`}>
            <thead>
                <tr>
                    <th scope="col">{_("Name")}</th>
                    <th scope="col" aria-label={_("Actions")} />
                </tr>
            </thead>
            <tbody>
                {tokens.map((token) => (
                    <TokenRow key={token.id} token={token} />
                ))}
            </tbody>
        </table>
    );
}

TokenRow.propTypes = {
    token: tokenShape,
};

function TokenRow({ token }) {
    return (
        <tr>
            <td className="align-middle">{token.name}</td>
            <td className="text-center">
                <TokenActions token={token} />
            </td>
        </tr>
    );
}

TokenActions.propTypes = {
    token: tokenShape,
};

function TokenActions({ token }) {
    const [setAlert] = useAlert();

    const [deleteTokenResponse, deleteToken] = useAPIDelete(
        `${API_URLs.tokens}/${token.id}`
    );
    useEffect(() => {
        if (deleteTokenResponse.state === API_STATE.ERROR) {
            setAlert(deleteTokenResponse.data);
        }
    }, [deleteTokenResponse, setAlert]);

    if (token.status === TOKEN_STATUS.REVOKED) {
        return <span className="text-muted">{_("Access revoked")}</span>;
    }
    if (
        token.status === TOKEN_STATUS.GENERATING ||
        deleteTokenResponse.state === API_STATE.SENDING
    ) {
        return <SpinnerElement />;
    }

    return (
        <div className="btn-group" role="group">
            <DownloadButton
                href={`${API_URLs.tokens}/${token.id}`}
                className="btn-primary btn-sm"
            >
                {_("Download")}
            </DownloadButton>
            <Button className="btn-danger btn-sm" onClick={deleteToken}>
                {_("Revoke")}
            </Button>
        </div>
    );
}

export default withErrorMessage(withSpinnerOnSending(TokensTable));
