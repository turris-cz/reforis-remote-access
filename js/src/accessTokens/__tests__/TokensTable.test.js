/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    render,
    getByText,
    fireEvent,
    wait,
} from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from "jest-mock-axios";

import { tokensFixture } from "./__fixtures__/tokens";
import TokensTable from "../TokensTable";

describe("<TokensTable />", () => {
    function renderTable(tokens) {
        return render(<TokensTable tokens={tokens} />);
    }

    function revokeToken(container) {
        fireEvent.click(getByText(container, "Revoke"));
    }

    it("should be rendered - empty table", () => {
        const { container } = renderTable([]);
        expect(container).toMatchSnapshot();
    });

    it("should be rendered - filled table", () => {
        const { container } = renderTable(tokensFixture);
        expect(container).toMatchSnapshot();
    });

    it("should display spinner while sending delete request", () => {
        const { container } = renderTable(tokensFixture);
        revokeToken(container);
        expect(mockAxios.delete).toBeCalledWith(
            "/reforis/remote-access/api/tokens/1",
            expect.anything()
        );
        expect(container).toMatchSnapshot();
    });

    it("should set alert on revoke error", async () => {
        const { container } = renderTable(tokensFixture);
        revokeToken(container);
        const errorMessage = "API didn't handle this well";
        mockJSONError(errorMessage);
        await wait(() => {
            expect(mockSetAlert).toHaveBeenCalledWith(errorMessage);
        });
    });
});
