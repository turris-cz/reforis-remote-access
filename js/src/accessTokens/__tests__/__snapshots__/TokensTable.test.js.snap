// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TokensTable /> should be rendered - empty table 1`] = `
<div>
  <p
    class="text-muted text-center"
  >
    There are no tokens added yet.
  </p>
</div>
`;

exports[`<TokensTable /> should be rendered - filled table 1`] = `
<div>
  <div
    class="table-responsive"
  >
    <table
      class="table table-hover"
    >
      <thead
        class="thead-light"
      >
        <tr>
          <th
            scope="col"
          >
            Name
          </th>
          <th
            aria-label="Actions"
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="align-middle"
          >
            foo
          </td>
          <td
            class="text-right"
          >
            <div
              class="btn-group"
              role="group"
            >
              <a
                class="btn btn-primary btn-sm"
                download=""
                href="/reforis/remote-access/api/tokens/1"
              >
                Download
              </a>
              <button
                class="btn btn-danger btn-sm"
                type="button"
              >
                Revoke
              </button>
            </div>
          </td>
        </tr>
        <tr>
          <td
            class="align-middle"
          >
            bar
          </td>
          <td
            class="text-right"
          >
            <span
              class="text-muted"
            >
              Access revoked
            </span>
          </td>
        </tr>
        <tr>
          <td
            class="align-middle"
          >
            baz
          </td>
          <td
            class="text-right"
          >
            <div
              class="spinner-border spinner-border-sm"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<TokensTable /> should display spinner while sending delete request 1`] = `
<div>
  <div
    class="table-responsive"
  >
    <table
      class="table table-hover"
    >
      <thead
        class="thead-light"
      >
        <tr>
          <th
            scope="col"
          >
            Name
          </th>
          <th
            aria-label="Actions"
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="align-middle"
          >
            foo
          </td>
          <td
            class="text-right"
          >
            <div
              class="spinner-border spinner-border-sm"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
        <tr>
          <td
            class="align-middle"
          >
            bar
          </td>
          <td
            class="text-right"
          >
            <span
              class="text-muted"
            >
              Access revoked
            </span>
          </td>
        </tr>
        <tr>
          <td
            class="align-middle"
          >
            baz
          </td>
          <td
            class="text-right"
          >
            <div
              class="spinner-border spinner-border-sm"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
